File: server/scripts/linkedin-capture.ts
/**
 * JobHawk — LinkedIn session capture script.
 *
 * Run this LOCALLY (not on Railway) once to capture your LinkedIn session:
 *
 *   cd server
 *   npx tsx scripts/linkedin-capture.ts
 *
 *   # Or point to your deployed Railway backend:
 *   npx tsx scripts/linkedin-capture.ts --url https://your-app.up.railway.app
 *
 * What it does:
 *   1. Opens a real Chrome window on your machine
 *   2. Navigates to LinkedIn login
 *   3. You log in manually (handles CAPTCHA, 2FA — it's your real browser)
 *   4. Captures your li_at session token
 *   5. Sends it to the backend → saved to data/linkedin-session.json
 *   6. LinkedIn will show as "Connected" next time you click Connect in the app
 *
 * The session lasts ~1 year. Re-run if LinkedIn shows "Session expired".
 */
import puppeteer from 'puppeteer-extra'
import StealthPlugin from 'puppeteer-extra-plugin-stealth'

puppeteer.use(StealthPlugin())

const cliArgs  = process.argv.slice(2)
const urlIdx   = cliArgs.indexOf('--url')
const BACKEND  = urlIdx !== -1 ? cliArgs[urlIdx + 1] : (process.env.BACKEND_URL ?? 'http://localhost:3001')

async function main(): Promise<void> {
  console.log('╔══════════════════════════════════════════════╗')
  console.log('║   JobHawk — LinkedIn Session Capture Tool   ║')
  console.log('╚══════════════════════════════════════════════╝')
  console.log(`\n  Backend: ${BACKEND}`)
  console.log('  Opening Chrome...\n')

  const browser = await puppeteer.launch({
    headless: false,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--window-size=1280,900'],
    defaultViewport: null,
  })

  const [page] = await browser.pages()
  await page.setExtraHTTPHeaders({ 'Accept-Language': 'en-US,en;q=0.9' })

  await page.goto('https://www.linkedin.com/login', {
    waitUntil: 'domcontentloaded',
    timeout: 30_000,
  })

  console.log('  ✋  Please log in to LinkedIn in the browser window.')
  console.log('      Waiting up to 5 minutes...\n')

  try {
    await page.waitForFunction(
      () => {
        const href = window.location.href
        return (
          href.includes('/feed') ||
          href.includes('/mynetwork') ||
          (href.includes('linkedin.com') && !href.includes('/login') && !href.includes('/checkpoint'))
        )
      },
      { timeout: 5 * 60_000, polling: 1000 },
    )
  } catch {
    console.error('\n  ❌  Timed out waiting for login. Run the script again.')
    await browser.close()
    process.exit(1)
  }

  console.log('  ✅  Login detected — capturing session...')
  // Let companion cookies settle
  await new Promise<void>((r) => setTimeout(r, 2500))

  const cookies = await page.cookies('https://www.linkedin.com', 'https://linkedin.com')
  const liAt    = cookies.find((c) => c.name === 'li_at')

  if (!liAt?.value) {
    console.error('\n  ❌  Could not find li_at cookie. The login may not have finished.')
    console.error('      Try waiting a few seconds after the feed loads, then run again.')
    await browser.close()
    process.exit(1)
  }

  // Try to extract the display name from the page
  let username = 'linkedin-user'
  try {
    const name = await page.evaluate(() => {
      const selectors = [
        '.feed-identity-module__member-name',
        '.profile-nav-card__name',
        '[data-control-name="nav.settings_signout"] ~ span',
        '.t-16.t-black.t-bold',
      ]
      for (const s of selectors) {
        const el = document.querySelector(s)
        if (el?.textContent?.trim()) return el.textContent.trim()
      }
      return null
    })
    if (name) username = name
  } catch { /* ignore — username is cosmetic */ }

  console.log(`\n  Token   : ${liAt.value.slice(0, 24)}...`)
  console.log(`  Username: ${username}`)
  console.log(`\n  Sending to ${BACKEND}...`)

  let result: { ok: boolean; error?: string }
  try {
    const res = await fetch(`${BACKEND}/api/auth/linkedin/import-session`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ liAt: liAt.value, username }),
    })
    result = await res.json() as typeof result
  } catch (err) {
    console.error(`\n  ❌  Could not reach backend at ${BACKEND}`)
    console.error(`      Make sure the server is running.`)
    console.error(`      Error: ${err instanceof Error ? err.message : String(err)}`)
    await browser.close()
    process.exit(1)
  }

  if (result.ok) {
    console.log('\n  ✅  Session saved!')
    console.log('      Click "Connect" for LinkedIn in the app — it will show as Connected.\n')
  } else {
    console.error(`\n  ❌  Backend error: ${result.error}`)
  }

  await browser.close()
}

main().catch((err) => {
  console.error('\n❌ Script crashed:', err instanceof Error ? err.message : String(err))
  process.exit(1)
})

File: server/scripts/linkedin_capture.py
#!/usr/bin/env python3
"""
JobHawk — LinkedIn Session Capture Script

Requirements:
    pip install selenium

Usage:
    python3 linkedin_capture.py
    python3 linkedin_capture.py --url https://your-railway-app.up.railway.app
"""

import sys
import time
import json
import urllib.request
import urllib.error


def check_selenium():
    try:
        import selenium  # noqa: F401
        return True
    except ImportError:
        print("\n  selenium is not installed.")
        print("  Run: pip install selenium")
        print("  Then re-run this script.\n")
        sys.exit(1)


def main():
    backend = "http://localhost:5173"
    #"https://jobhawk-server-production.up.railway.app"
    args = sys.argv[1:]

    if "--url" in args:
        idx = args.index("--url")
        if idx + 1 < len(args):
            backend = args[idx + 1].rstrip("/")

    print("╔══════════════════════════════════════════════╗")
    print("║   JobHawk — LinkedIn Session Capture Tool   ║")
    print("╚══════════════════════════════════════════════╝")
    print(f"\n  Backend: {backend}")

    check_selenium()

    from selenium import webdriver
    from selenium.webdriver.chrome.options import Options

    print("  Opening Chrome...\n")

    options = Options()
    options.add_argument("--window-size=1280,900")

    try:
        driver = webdriver.Chrome(options=options)
    except Exception as e:
        print(f"\n  Could not open Chrome: {e}")
        print("  Make sure Google Chrome is installed on your system.")
        sys.exit(1)

    try:
        driver.get("https://www.linkedin.com/login")

        print("  Please log in to LinkedIn in the browser window.")
        print("  Waiting up to 5 minutes...\n")

        timeout = 5 * 60
        start = time.time()
        logged_in = False

        while time.time() - start < timeout:
            try:
                url = driver.current_url
            except Exception:
                break

            if (
                "/feed" in url
                or "/mynetwork" in url
                or (
                    "linkedin.com" in url
                    and "/login" not in url
                    and "/checkpoint" not in url
                    and "/uas/" not in url
                )
            ):
                logged_in = True
                break

            time.sleep(1)

        if not logged_in:
            print("\n  Timed out waiting for login. Run the script again.")
            sys.exit(1)

        print("  Login detected — capturing session...")
        time.sleep(2.5)

        cookies = driver.get_cookies()
        li_at_cookie = next((c for c in cookies if c["name"] == "li_at"), None)

        if not li_at_cookie:
            print("\n  Could not find li_at cookie.")
            print("  The login may not have fully completed.")
            print("  Try waiting a few seconds on the feed, then run again.")
            sys.exit(1)

        token = li_at_cookie["value"]

        # Try to extract the display name from the page
        username = "linkedin-user"
        for selector in [
            ".feed-identity-module__member-name",
            ".profile-nav-card__name",
            ".t-16.t-black.t-bold",
        ]:
            try:
                elements = driver.find_elements("css selector", selector)
                if elements and elements[0].text.strip():
                    username = elements[0].text.strip()
                    break
            except Exception:
                pass

        print(f"\n  Token   : {token[:24]}...")
        print(f"  Username: {username}")
        print(f"\n  Sending to {backend}...")

        payload = json.dumps({"liAt": token, "username": username}).encode("utf-8")
        req = urllib.request.Request(
            f"{backend}/api/auth/linkedin/import-session",
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )

        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                result = json.loads(resp.read().decode("utf-8"))
        except urllib.error.URLError as e:
            print(f"\n  Could not reach backend at {backend}")
            print("  Make sure the server is running.")
            print(f"  Error: {e}")
            sys.exit(1)

        if result.get("ok"):
            print("\n  Session saved!")
            print("  Click 'Connect' for LinkedIn in the app — it will show as Connected.\n")
        else:
            print(f"\n  Backend error: {result.get('error', 'Unknown error')}")

    finally:
        time.sleep(2)
        try:
            driver.quit()
        except Exception:
            pass


if __name__ == "__main__":
    main()
